pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go

```diff
@@ -2893,3 +2921,220 @@ func TestGetIntOrPercentValue(t *testing.T) {
+		spyClient.AssertNumberOfCalls(t, "Create", 0)
+		spyClient.AssertNumberOfCalls(t, "Delete", 0)
+	}
+}
+func newMachineHealthCheckWithRemediationTemplate(infraRemediationTmpl *unstructured.Unstructured) *mapiv1beta1.MachineHealthCheck {
+
+	mhc := maotesting.NewMachineHealthCheck("machineHealthCheck")
+	remediationTemplateObjRef := &corev1.ObjectReference{
+		APIVersion: "infrastructure.machine.openshift.io/v1alpha3",
+		Kind:       "InfrastructureRemediationTemplate",
+		Name:       infraRemediationTmpl.GetName(),
+	}
+
+	mhc.Spec.RemediationTemplate = remediationTemplateObjRef
+	return mhc
+}
+
+func newCreateRequest(ownerMachine *mapiv1beta1.Machine) *unstructured.Unstructured {
+	m := maotesting.NewExternalRemediationMachine()
+	md := m.Object["metadata"].(map[string]interface{})
+	md["annotations"] = map[string]interface{}{
+		"machine.openshift.io/cloned-from-name":      "",
+		"machine.openshift.io/cloned-from-groupkind": "InfrastructureRemediationTemplate.infrastructure.machine.openshift.io",
+	}
+	orElement := map[string]interface{}{
+		"apiVersion": ownerMachine.APIVersion,
+		"kind":       ownerMachine.Kind,
+		"name":       ownerMachine.Name,
+		"uid":        ownerMachine.UID,
+	}
+	md["ownerReferences"] = []interface{}{orElement}
+	md["resourceVersion"] = "1"
+	return m
+}
+
+//Matchers
+func matchGetERM() interface{} {
+	f := func(actual interface{}) bool {
+		actualRemediationQuery, ok := actual.(*unstructured.Unstructured)
+		if !ok {
+			return false
+		}
+		md, ok := actualRemediationQuery.Object["metadata"].(map[string]interface{})
+		if !ok {
+			return false
+		}
+		return actualRemediationQuery.Object["kind"] == "InfrastructureRemediation" &&
+			actualRemediationQuery.Object["apiVersion"] == "infrastructure.machine.openshift.io/v1alpha3" &&
+			md["name"] == "Machine"
+	}
+	return mock.MatchedBy(f)
+}
+
+func buildMatcher(expected interface{}) interface{} {
+	f := func(actual interface{}) bool {
+
+		if actual == nil || expected == nil {
+			return actual == expected
+		}
+
+		if reflect.TypeOf(expected) != reflect.TypeOf(actual) {
+			return false
+		}
+
+		return fmt.Sprint(actual) == fmt.Sprint(expected)
+	}
+
+	return mock.MatchedBy(f)
+}
+
+func (cw *clientWrap) Create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error {
+	cw.Mock.On("Create", ctx, obj, opts).Return(nil)
+	cw.Mock.Called(ctx, obj, opts)
+	return cw.Client.Create(ctx, obj, opts...)
+}
+
+func (cw *clientWrap) Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+	cw.Mock.On("Delete", ctx, obj, opts).Return(nil)
+	cw.Mock.Called(ctx, obj, opts)
+	return cw.Client.Delete(ctx, obj, opts...)
+}
+
+func (cw *clientWrap) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	cw.Mock.On("Get", ctx, key, obj).Return(nil)
+	cw.Mock.Called(ctx, key, obj)
+	return cw.Client.Get(ctx, key, obj)
+}
```

Comment:
I'm not a fan of using spy clients like this, normally we would expect to just look up the object after the code under test has run. It's more explicit and helps us understand exactly what the input/output is.

Eg, a test should look something like this
```
// Create fake client
// Run remediation code
// Use fake client to check external remediation object now exists
// Check external remediation object matches the expectation 
```

Reply:
I must disagree here.
In the pattern you suggest I don't like the double invocation of the client.
However I am a guest in your repo :smile: , so let me know if it something you feel strongly about - if so I'll try to change the pattern.

Reply:
To each their own I guess, to me it makes sense to call the client to see what it thinks the state is (we aren't testing the client here, so we trust the fake to give us the right result)

I'd prefer we keep to the existing patterns in the repo ultimately, so there's a lower burden for maintenance of the tests in the future

Reply:
NP, I've made the changes :+1: